core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -348,10 +348,24 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   // Thread Local variable that can be used by users to pass information down the stack
   private val localProperties = new InheritableThreadLocal[Properties] {
-    override protected def childValue(parent: Properties): Properties = new Properties(parent)
+    override protected def childValue(parent: Properties): Properties = {
+      // Note: make a clone such that changes in the parent properties aren't reflected in
+      // the those of the children threads, which has confusing semantics (SPARK-10564).
+      val p = new Properties
+      val filtered = parent.asScala.filter { case (k, _) =>
+        !nonInheritedLocalProperties.contains(k)
+      }
+      p.putAll(filtered.asJava)
+      p
+    }
     override protected def initialValue(): Properties = new Properties()
   }
 
+  /**
+   * Keys of local properties that should not be inherited by children threads.
+   */
+  private[spark] val nonInheritedLocalProperties: HashSet[String] = new HashSet[String]
```

Comment:
the whole point of SPARK-10548 is to avoid inheriting the SQL execution ID. How can we avoid that with just cloning?